### ISSUE TITLE
Swarm mode mixing requests from unrelated services

### ISSUE THREAD

**Issue by dangra** [docker/docker#26590]
**Description**

Running 2 services (swarm mode) whose containers listen on the same internal port confuses networking layer and mix client requests.

For example running below webserver (Also found as https://github.com/dangra/swarm-net-bug) 

```go
package main

import (
	"flag"
	"fmt"
	"net/http"
)

var expectedid = flag.String("id", "", "An unique id")

func handler(w http.ResponseWriter, r *http.Request) {
	gotid := r.URL.Path[1:]
	if *expectedid != gotid {
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprintf(w, "BAD -- Expected '%s', Got '%s'\n", *expectedid, gotid)
	} else {
		fmt.Fprintf(w, "OK\n")
	}
}

func main() {
	flag.Parse()
	http.HandleFunc("/", handler)
	http.ListenAndServe(":8080", nil)
}
```

**Steps to reproduce the issue:**

1. Clone and build image:
    `git clone https://github.com/dangra/swarm-net-bug`
    `docker build -t swarm-net-bug swarm-net-bug`

2. Then create 2 services on published port 10001 and 10002: 
    `docker service create --name s1 -p 10001:8080 swarm-net-bug app --id 1`
    `docker service create --name s2 -p 10002:8080 swarm-net-bug app --id 2`

3. Finally use 2 different terminals to query both services in a loop:
    `while true; do curl http://localhost:10001/1; done`
    `while true; do curl http://localhost:10002/2; done`

**Describe the results you received:**

After a while you will see http responses from the service you are not requesting for:

```
...
OK
OK
OK
OK
OK
OK
OK
OK
BAD -- Expected '1', Got '2'
OK
OK
BAD -- Expected '1', Got '2'
OK
OK
OK
OK
...
```   

**Describe the results you expected:**

I expect all requests to service `s1` to reach only containers from the same service and not others.
In summary, not a single _"BAD - Expected A, got B"_ line of output.

**Additional information you deem important (e.g. issue happens only occasionally):**

I reproduced the issue on a clean setup on Ubuntu 14.04 and also in Docker for Mac.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 21
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay null host
Swarm: active
 NodeID: cr5rln1hk70p3iofosers97xj
 Is Manager: true
 ClusterID: 6wli6ivsoslm6lb3n5gkvr0hq
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 136.243.74.78
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.19.0-68-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 62.82 GiB
Name: w01.swarm01.xxxxxxxxxxxx.com
ID: W4NN:TW4G:L4QA:N6BM:JMJO:JTAO:6NTV:OZB4:7GVH:TK3R:PR43:V3SN
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Comment by mrjana** [comment]
@dangra Thanks for a simple repro. I was able to reproduce the issue. On initial analysis it looks like a problem of the client using the same source port before the connection expires from ipvs connection table (which happens after the connection goes out of FIN_WAIT state). But need to analyze a bit more to see if this is a kernel issue or an issue on docker side.

**Comment by mrjana** [comment]
@dangra More analysis indicates that there might be a kernel issue at play here but since this happens in such an atypical case that nobody else using ipvs would have encountered this. What happens is as I mentioned the client starts reusing the same source port (it could be used to reach `s2` in your test code while the previous connection may have used the same source port for reaching `s1`) before ipvs connection table clears that flow. 

When that happens the incoming new connection request is matched against this old/stale entry which hasn't been removed from ipvs because we share everything between this flow and the previous flow i.e the source port is matching, source IP and dest IP are in fact the same for connections to both services. and dest port is also the same i.e `8080` and when this happens we route to the wrong service.

To confirm this is what is happening I completely removed the REDIRECT rule from ingress sandbox so that dest port does not get changed from 10001/10002 to 8080 and added the same REDIRECT rule in the destination container and the problem disappeared because after this change the flows going to these two different services will no longer collide even transiently.

This mainly happens because of a constant while loop of connection setup/teardown.

**Comment by dangra** [comment]
> but since this happens in such an atypical case that nobody else using ipvs would have encountered this.

@mrjana not sure I follow why the case is atypical. To me is quite common to have a load balancer and/or reverse proxy that routes requests to different websites (swarm services) based on Host header.

> To confirm this is what is happening I completely removed the REDIRECT rule from ingress sandbox so that dest port does not get changed from 10001/10002 to 8080 and added the same REDIRECT rule in the destination container and the problem disappeared because after this change the flows going to these two different services will no longer collide even transiently.

It makes sense.So, do you confirm it is a problematic bug and can be fixed in Docker by moving the REDIRECT to container network namespace?

**Comment by aaronlehmann** [comment]
@mrjana: Do you think the fix should be backported to 1.12.2?

**Comment by mrjana** [comment]
Yes, this will most likely make it to the patch release.
